### PR TITLE
Escher feature/vamm get twap

### PR DESCRIPTION
**Pull request by Cardosaum**
## Issue
* [CU-24ah03q](https://app.clickup.com/t/24ah03q)

## Description
* Add `get_twap` functionality to the Vamm Pallet

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-24ah03q Feature 6 (Get TWAP Functionality)](https://app.clickup.com/t/24ah03q)

File: frame/vamm/src/lib.rs [hunk 1]
```diff
@@ -259,6 +264,18 @@ pub mod pallet {
...
+		pub base_asset_twap: Balance,
+
+		// TODO(Cardosaum): Write description for this field.
+		pub base_asset_twap_timestamp: Moment,
+
+		// TODO(Cardosaum): Write description for this field.
+		pub quote_asset_twap: Balance,
+
+		// TODO(Cardosaum): Write description for this field.
+		pub quote_asset_twap_timestamp: Moment,
```
Comment (0xangelo) [inline frame/vamm/src/lib.rs:278]: Are these updated anywhere?

File: frame/vamm/src/tests/get_twap.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			vamms: vec![(0, vamm_state)]
+		}.build().execute_with(|| {
+			// In order to check if vamm is closed or not we must simulate time
+			// passing.
+			run_for_seconds(
+				if let Some(t) = vamm_state.closed {
+					t + 1
+				} else {
+					1
+				}
```
Comment (0xangelo) [inline frame/vamm/src/tests/get_twap.rs:63]: Maybe a `match` statement would be shorter?
```rust
match vamm_state.closed { Some(t) => t + 1, _ => 1 }
```

**Review by 0xangelo — COMMENTED** [review]
Looking good. I like the documentation improvements and refactorings. I'm guessing the logic for updating `base/quote_asset_twap` will be implemented later?

File: frame/vamm/src/lib.rs (hunk 1 above)
Comment (Cardosaum) [reply frame/vamm/src/lib.rs:278]: Yeah they will, I just opted to open a new PR for it. :)

**Review by 0xangelo — APPROVED** [review]

File: frame/vamm/src/lib.rs [hunk 3]
```diff
@@ -338,12 +357,14 @@ pub mod pallet {
...
 		/// computation was not successful.
 		FailedToDeriveInvariantFromBaseAndQuoteAsset,
-		/// Tried to perform swap operation but it would drain all base asset reserves.
+		/// Tried to perform swap operation but it would drain all [`base`](VammState::base_asset_reserves) asset reserves.
 		BaseAssetReservesWouldBeCompletelyDrained,
-		/// Tried to perform swap operation but it would drain all quote asset reserves.
+		/// Tried to perform swap operation but it would drain all
+		/// [`quote`](VammState::quote_asset_reserves) asset reserves.
 		QuoteAssetReservesWouldBeCompletelyDrained,
 	}
```
Comment (kevin-composable) [inline frame/vamm/src/lib.rs:369]: This isn't required but you could provide links to all functions that can result in each error

File: frame/vamm/src/lib.rs [hunk 4]
```diff
@@ -529,9 +561,60 @@ pub mod pallet {
...
+		/// ## Parameters
+		///  - [`vamm_id`](Config::VammId): The ID of the desired vamm to query.
+		///  - [`asset_type`](composable_traits::vamm::AssetType): The desired
+		///  asset type to get info about.
+		///
+		/// ## Returns
+		/// The twap for the specified asset.
+		///
+		/// ## Assumptions or Requirements
+		/// * The requested [`VammId`](Config::VammId) must exists
```
Comment (kevin-composable) [inline frame/vamm/src/lib.rs:583]: exists -> exist. You can add this to another PR tho so you don't have to redo this one

**Review by kevin-composable — APPROVED** [review]

**Comment by 0xangelo** [comment]
@Cardosaum can you check that changes don't conflict with Clearing House before merging?

File: frame/vamm/src/lib.rs (hunk 3 above)
Comment (Cardosaum) [reply frame/vamm/src/lib.rs:369]: Oh, okay!

File: frame/vamm/src/lib.rs (hunk 4 above)
Comment (Cardosaum) [reply frame/vamm/src/lib.rs:583]: Thanks! :)

**Comment by Cardosaum** [comment]
> @Cardosaum can you check that changes don't conflict with Clearing House before merging?

I made modifications in the Clearing House (only two places needed), and now the code is compiling just fine and passing all tests. :+1:

**Pull request merged**